include/zephyr/device.h

```diff
@@ -403,6 +403,28 @@ typedef int16_t device_handle_t;
  */
 #define DEVICE_DECLARE(name) static const struct device DEVICE_NAME_GET(name)
 
+/**
+ * @def DEVICE_INIT_DT_GET
+ *
+ * @brief Get a <tt>const struct init_entry*</tt> from a devicetree node
+ *
+ * @param node_id A devicetree node identifier
+ *
+ * @return A pointer to the init_entry object created for that node
+ */
+#define DEVICE_INIT_DT_GET(node_id) (&Z_INIT_ENTRY_NAME(DEVICE_DT_NAME_GET(node_id)))
+
+/**
+ * @def DEVICE_INIT_GET
+ *
+ * @brief Get a <tt>const struct init_entry*</tt> from a device by name
+ *
+ * @param name The same as dev_name provided to DEVICE_DEFINE()
+ *
+ * @return A pointer to the init_entry object created for that device
+ */
+#define DEVICE_INIT_GET(name) (&Z_INIT_ENTRY_NAME(DEVICE_NAME_GET(name)))
```

Comment:
is this something applications want to use? Just saying because these may be internal APIs instead.

Reply:
I believe so, the purpose of these macros is for injecting dependencies into devices and SYS_INIT macros.
As dependencies can be devices or other SYS_INIT macros, the only common element they have is the init struct.
I see the solution being something like this:
```
SYS_INIT(some_functionality, APPLICATION, 10,
    /*  Dependencies */
    SYS_INIT_GET(kernel),
    SYS_INIT_GET(clock),
    DEVICE_INIT_GET(DT_ALIAS(button0)));
```

Reply:
I see, thanks. To be honest, this way of indicating dependencies feels a bit weird. I'd like to set a dependency on sus_init/device, not based on init related objects. The code above is not super clear for people not familiar with Zephyr internals. I know these are dependencies because there's a comment, but who would guess that by just reading the last 3 lines?

But ultimately, I see the point, this enables functionality by doing minimal changes, which is good. However, I'd improve this area before extending this pattern everywhere. For example, `SYS_INIT` could be replaced by the concept of system services (and that `struct device *` in each sys init call removed....). Then you'd have services and devices, for which you could specify dependencies, maybe using a set of macros that make things clearer to the user.

Reply:
To make things clearer/more explicit it could be more like:
```
SYS_INIT(some_functionality, APPLICATION, 10)
SYS_INIT_DEPS(some_functionality,
    SYS_INIT_GET(kernel),
    SYS_INIT_GET(clock),
    DEVICE_INIT_GET(DT_ALIAS(button0)));
```
And the same equivalent for devices. The list would just be creating some array that is thrown away after the first linker pass anyway, it doesn't need to be part of the main macro for any reason.

Reply:
I think that would be more readable, definitely.